Add ufuncs

AaronDJohnson opened the pull request:
Add ufuncs for the following operations from the `QD` [library:](https://github.com/scibuilder/QD)

* `arctan2`
* `arctan`
* `arccos`
* `arcsin`
* `arctanh`
* `arccosh`
* `arcsinh`
* `power`
* `modf`
* `ldexp`

Tests have also been added for each of these new functions.

AaronDJohnson committed:
add ldexp

AaronDJohnson committed:
remove some unnecessary bits

AaronDJohnson committed:
add tan and modify kind

kind is now 'f'.... This is supposedly bad, but makes my job easier for now.

AaronDJohnson committed:
add atan

This still needs to be tested.

AaronDJohnson committed:
add missing ufuncs

Several ufuncs have been added in this commit:

arctrig functions
power
fmod
modf <--- to be tested

AaronDJohnson committed:
Merge remote-tracking branch 'upstream/mainline' into for_pint

AaronDJohnson committed:
fix order of output for modf function

AaronDJohnson committed:
change back to void type for PR

AaronDJohnson committed:
fix arctan2, add tests

AaronDJohnson committed:
remove fmod

This function doesn't exist in QD and my implementation didn't work properly.

AaronDJohnson commented:
This should fix #10. Note that there is some support for negative `a` values in `numpy` with `np.power(a, b)` which can be raised to integer values. This coverage is not replicated in the `QD` functions.

AaronDJohnson commented:
I'm not sure what is going on with the `ubuntu` test for `arcsin` and `arccos`. These pass locally on my M1 Mac. I'll try them on my Ubuntu computer and see if they fail locally too.

AaronDJohnson commented:
It does indeed fail on my Linux computer. I'll probably need some help diagnosing the issue. Both `arcsin` and `arccos` are failing.

A few things that I have tried:
* I computed the function definition for both `arcsin` and `arccos`: e.g. `atan2qq(a, sqrtq(subdq(1.0, sqrq(a))))` using their numpy ufuncs on `ddouble`s: `np.atan2qq(a, np.sqrt(1 - np.sqr(a)))`. This returned results in agreement with the numpy result 
* Using `x = np.linspace(-1, 1, 100)` and `x_d = x.astype(ddouble)`, I find `np.arcsin(x) - np.arcsin(x_d)` is at most ~1e-16. However, the test, which I think does the same thing, fails.
* Doing the same for `arccos` does show some obvious disagreement. However, picking the pieces apart does not show any obvious disagreement.

I hope I'm missing something obvious here!

mwallerb commented on csrc/dd_arith.c:
```diff
@@ -488,3 +494,244 @@ ddouble tanhq(ddouble a)
+}
+
+ddouble powdq(const double a, const ddouble b) {
+    powqq(adddq(a, Q_ZERO), b);
```
Careful, you forgot the return!
```suggestion
    return powqq((ddouble){a, 0.0}, b);
```

AaronDJohnson replied:
Nice catch! Since there is not much code here, I've changed it to deal with the `double` type directly instead of going through `powqq`.

mwallerb commented on csrc/dd_arith.c:
```diff
@@ -488,3 +494,244 @@ ddouble tanhq(ddouble a)
+    if (equalqq(a, Q_ZERO) && equalqq(b, Q_ZERO)) {
+        return Q_ONE;
+    } else if (equalqq(a, Q_ZERO) && !equalqq(b, Q_ZERO)) {
+        return Q_ZERO;
```
Suggest using `iszeroq` here, which should be slightly faster

```suggestion
    if (iszeroq(a) && iszeroq(b)) {
        return Q_ONE;
    } else if (iszeroq(a) && !iszeroq(b)) {
        return Q_ZERO;
```

AaronDJohnson replied:
Fixed!

mwallerb commented on csrc/dd_arith.c:
```diff
@@ -488,3 +494,244 @@ ddouble tanhq(ddouble a)
+ddouble atanhq(const ddouble a) {
+    if (equalqd(a, -1.0)) {
+        return negq(infq());
+    } else if (equalqd(a, 1.0)) {
```
Suggest use isone here
```suggestion
    } else if (isoneq(a)) {
```

AaronDJohnson replied:
Fixed!

mwallerb commented on csrc/dd_arith.c:
```diff
@@ -488,3 +494,244 @@ ddouble tanhq(ddouble a)
+        denominator is larger.  Otherwise, the second is used.
+    */
+
+    if (iszeroq(x)) {
+
+        if (iszeroq(y)) {
+            /* Both x and y is zero. */
+            return Q_ZERO;
+    }
+
```
This looks a little messy
```suggestion
    if (iszeroq(x)) {
        if (iszeroq(y))
            return Q_ZERO;           /* Both x and y are zero */
```

AaronDJohnson replied:
Fixed

mwallerb commented on csrc/dd_arith.c:
```diff
@@ -488,3 +494,244 @@ ddouble tanhq(ddouble a)
+    ddouble yy = divqq(y, r);
+
+    /* Compute double precision approximation to atan. */
+    // ddouble z = (ddouble){atan2(y.hi, x.hi), 0.};
+    ddouble z = adddq(atan2(y.hi, x.hi), Q_ZERO);
```
Suggest going back to the commented out version, as it should be faster IMHO.
```suggestion
    ddouble z = (ddouble){atan2(y.hi, x.hi), 0.};
```

AaronDJohnson replied:
I've changed this here and in a couple of other places that I did a similar thing.

mwallerb commented on csrc/dd_arith.h:
```diff
@@ -506,11 +508,34 @@ static inline ddouble hypotqd(ddouble x, double y)
 }
 
+/* Computes the nearest integer to d. */
+inline ddouble nintq(ddouble d) {
```
This needs to be static inline, because of C weird inline rules
```suggestion
static inline ddouble nintq(ddouble d) {
```

AaronDJohnson replied:
Static has been added.

mwallerb commented on csrc/dd_arith.c:
```diff
@@ -181,6 +182,11 @@ ddouble expm1q(ddouble a)
     return subqd(sum, 1.0);
 }
 
+ddouble inline ldexpqi(ddouble a, int exp)
```
You need to remove inline here ... C is very peculiar.
```suggestion
ddouble ldexpqi(ddouble a, int exp)
```

AaronDJohnson replied:
`inline` removed

mwallerb commented on csrc/dd_arith.c:
```diff
@@ -488,3 +494,244 @@ ddouble tanhq(ddouble a)
+    q = floor(t.hi / Q_PI_16.hi + 0.5);
+    t = subqq(t, mulqd(Q_PI_16, q));
+    int k = (int)(q);
+    int abs_k = (int)(fabs(k));
```
I think this should just be `abs`, since `k` is integer? Or is there something wrong with the logic?

AaronDJohnson replied:
It definitely should be `abs` instead of `fabs`. Oops! This has been fixed in the recent commits.

mwallerb reviewed (commented):
Dear @AaronDJohnson,

thanks *so much* for this!  This PR already looks great!

As for the failing test: this is **very** strange - it seems `atan2()` somehow flips the arguments around, which I cannot reproduce ... maybe we can fix this after merging too ...

mwallerb commented:
Something weird is going on at 45 degrees (1/sqrt 2)...

![image](https://user-images.githubusercontent.com/22056626/169802020-a59affda-798c-459a-9cf1-dd286dd7961e.png)

AaronDJohnson committed:
recommended changes

equalqq(x, Q_ONE) changed to isoneq(x)

equalqq(x, Q_ZERO) changed to iszeroq(x)

powdq works on double types directly instead of going through powqq

use type casting to ddouble type instead of adding zero to do the casting

fix static/inline in a couple of places

fabs -> abs for integers

AaronDJohnson committed:
clean up atan2qd/atan2dq

mwallerb commented on csrc/dd_arith.c:
```diff
@@ -488,3 +494,245 @@ ddouble tanhq(ddouble a)
+    return divqq(s, c);
+}
+
+void sincos(const ddouble a, ddouble *sin_a, ddouble *cos_a) {
```
@AaronDJohnson  I think I found the bug!

Can you rename this function `sincosq` and change the call site below?

AaronDJohnson replied:
This fixes it locally on my computer with Pop!OS. It looks like the tests would pass with ulps=2. This is still a bit strange to me though, because ulps=1 works for a similar environment for me.

How did you find this bug? Nice work!

AaronDJohnson committed:
rename sincos to sincosq

AaronDJohnson committed:
change arcsin ulps to 2
